### PR TITLE
[LoRA] text encoder: read the ranks for all the attn modules

### PR DESCRIPTION
**Pull request by elias-gaeros**
# What does this PR do?

The enumeration of LoRA adapters for gathering their ranks in `LoraLoaderMixin.load_lora_into_text_encoder` is modified in order to: 

 * In addition to out_proj, read the ranks of adapters for q_proj, k_proj, and  v_proj

 * Allow missing adapters (UNet already supports this)

# Motivation

Some resized LoRAs have a different rank for each adapter module. Currently, only the ranks of the `out_proj`  attention modules are gathered in the [`rank` dictionary](https://github.com/huggingface/diffusers/blob/5bb3858/src/diffusers/loaders/lora.py#L569-L570). Using a `LoraConfig` with missing `rank_patterns`  result in the following errors:
```
RuntimeError: Error(s) in loading state_dict for CLIPTextModel:
        size mismatch for text_model.encoder.layers.0.self_attn.k_proj.lora_A.default_0.weight: copying a param with shape torch.Size([8, 768]) from checkpoint, the shape in current model is torch.Size([3, 768]).
        size mismatch for text_model.encoder.layers.0.self_attn.k_proj.lora_B.default_0.weight: copying a param with shape torch.Size([768, 8]) from checkpoint, the shape in current model is torch.Size([768, 3]).
```

UNet supports missing adapter modules, but the text encoder doesn't, resulting in:
```
  File "/home/elias/repos/diffusers/src/diffusers/loaders/lora.py", line 1363, in load_lora_weights
    self.load_lora_into_text_encoder(
  File "/home/elias/repos/diffusers/src/diffusers/loaders/lora.py", line 572, in load_lora_into_text_encoder
    rank[rank_key] = text_encoder_lora_state_dict[rank_key].shape[1]
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^
KeyError: 'text_model.encoder.layers.11.self_attn.out_proj.lora_B.weight'
```

`encoder.layers.11.*.lora_B.weight` is often all zeroes when the layer is skipped during fine-tuning, in this case it might be desirable to not store them in LoRA safetensors.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? **I could'nt find an open issue or forum thread about the problems**
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sayakpaul is the author of the lines affected by this PR.

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks for your contributions!

Could you also add a test to ensure the robustness of the changes?

**Review by younesbelkada — APPROVED** [review]
Clean, thanks !

**Comment by elias-gaeros** [comment]
> Thanks for your contributions!
> 
> Could you also add a test to ensure the robustness of the changes?

@sayakpaul I may need some guidance for writing the tests. Testing this change requires a specially resized LoRA. 
What about:
 * Using [`test_sdxl_1_0_lora`](https://github.com/huggingface/diffusers/blob/5bb3858/tests/lora/test_lora_layers_sdxl.py#L183) as a template,
 * Making a PR to [hf-internal-testing/sdxl-1.0-lora](https://huggingface.co/hf-internal-testing/sdxl-1.0-lora) with `sd_xl_offset_example-resized-lora_1.0.safetensors`? or to different repository?
 
 Alternatively the test could drop layers and ranks in the state_dict, but the output will be nonsensical without orthogonalization.

**Comment by sayakpaul** [comment]
Sure. 

> @sayakpaul I may need some guidance for writing the tests. Testing this change requires a specially resized LoRA.

So, we could add a utility to create a `peft` LoRA config that tests this use case specifically. And then add an actual test case for it. We could add the utility and the test case to [this file](https://github.com/huggingface/diffusers/blob/main/tests/lora/utils.py).

WDYT?

File: tests/lora/utils.py
Comment (sayakpaul) [inline tests/lora/utils.py:446]: I think we should keep this behavior truly deterministic, i.e., don't rely on `rng.choice()` and just manually remove the adapters. I prefer this to be a slightly better for testing.

Comment (sayakpaul) [reply tests/lora/utils.py:446]: I also think that the test should:

* Run a single inference pass after the LoRAs have been added to the text encoder. Then we compare the outputs to the non-LoRA case. 
* Run another inference pass with adapters discarded and compare the outputs to the original LoRA outputs. 
* We should do the same on the UNet, too.

**Review by sayakpaul — COMMENTED** [review]
Thanks for the tests. I left some comments.

File: tests/lora/utils.py
Comment (elias-gaeros) [reply tests/lora/utils.py]: I updated the test, now removing all adapters from `text_model.encoder.layers.4`.

File: tests/lora/utils.py
Comment (sayakpaul) [inline tests/lora/utils.py:427]: Would be nice to accompany this with a comment to explain its existence.

**Review by sayakpaul — APPROVED** [review]
Just a minor comment. Rest looks really good to me. Thanks for taking care of it!

**Comment by sayakpaul** [comment]
Thanks so much for the iterations! Will merge once the CI is green.

**Comment by elias-gaeros** [comment]
What is keeping this from being merged?

**Comment by sayakpaul** [comment]
That will be on me. I forgot to merge it. Let the CI run once again and I will merge.

**Comment by sayakpaul** [comment]
Ah looks like need to run a couple of quality related formatting. Keeping it open and will take care of it once I am back to my keyboard.

**Comment by sayakpaul** [comment]
I am unable to locally reproduce the LoRA related test failures :/ Seems like some NumPy version mismatch bug. Can quickly open a PR to pin the NumPy version to be down under 2 so that we can unblock the PRs. @yiyixuxu @DN6 WDYT?

**Comment by sayakpaul** [comment]
The failing test is completely unrelated and sorry for the delay on my end.

**Pull request merged**
